content/docs/04-ai-sdk-rsc/03-generative-ui-state.mdx

```diff
@@ -0,0 +1,268 @@
+---
+title: Managing Generative UI State
+description: Overview of the AI and UI states
+---
+
+# Managing Generative UI State
+
+State is an essential part of any application. State is particularly important in AI applications as it is passed to large language models (LLMs) on each request to ensure it has the necessary context to produce a great generation. Traditional chatbots are text-based and have a structure that mirrors that of any chat application.
+
+For example, in a chatbot, state is an array of `messages` where each `message` has:
+
+- `id`: a unique identifier
+- `role`: who sent the message (user/assistant/system/tool)
+- `content`: the content of the message
+
+This state can be rendered in the UI and sent to the model without any modifications.
+
+With Generative UI, the model is now returning a React component, rather than a plain text message. The client can render that component without issue, but that state can't be sent back to the model because React components aren't serialisable. So, what can you do?
+
+**The solution is to split the state in two, where one (AI State) becomes a proxy for the other (UI State)**.
+
+One way to understand this concept is through a Lego analogy. Imagine a 10,000 piece Lego model that, once built, cannot be easily transported because it is fragile. By taking the model apart, it can be easily transported, and then rebuilt following the steps outlined in the instructions pamphlet. In this way, the instructions pamphlet is a proxy to the physical structure. Similarly, AI State provides a serialisable (JSON) representation of your UI can be passed back and forth to the model.
```

Comment:
```suggestion
One way to understand this concept is through a Lego analogy. Imagine a 10,000 piece Lego model that, once built, cannot be easily transported because it is fragile. By taking the model apart, it can be easily transported, and then rebuilt following the steps outlined in the instructions pamphlet. In this way, the instructions pamphlet is a proxy to the physical structure. Similarly, AI State provides a serialisable (JSON) representation of your UI that can be passed back and forth to the model.
```